Build SOF with Zephyr on i.MX

iuliana-prodan opened the pull request:
Add support for i.MX to build Sound Open Firmware with Zephyr OS on DSP.
 The build target is i.MX8QM. It has one Xtensa HiFi4 core, with 64 KB TCM,
448 KB OCRAM, 8MB SDRAM and 1 ESAI, 1 SAI as audio interfaces.

Please check each commit message.

With these commits I have a successful build.
I obtain the zephyr.ri image which boots the DSP and loads the topology.

```
root@imx8qmmek:~# ./loadSof.sh 
[  166.405965] sof-audio-of 556e8000.dsp: DT DSP detected
[  166.416971] imx-dsp imx-dsp: NXP i.MX DSP IPC initialized
[  166.430784] sof-audio-of 556e8000.dsp: warning: unknown sof_ext_man header type 5 size 0x20
[  166.439213] sof-audio-of 556e8000.dsp: warning: unknown sof_ext_man header type 3 size 0x30
[  166.447885] sof-audio-of 556e8000.dsp: Firmware info: version 1:8:0-cf919
[  166.454703] sof-audio-of 556e8000.dsp: Firmware: ABI 3:19:0 Kernel ABI 3:17:0
[  166.461862] sof-audio-of 556e8000.dsp: warn: FW ABI is more recent than kernel
[  166.469453] sof-audio-of 556e8000.dsp: Firmware info: version 1:8:0-cf919
[  166.476264] sof-audio-of 556e8000.dsp: Firmware: ABI 3:19:0 Kernel ABI 3:17:0
[  166.483406] sof-audio-of 556e8000.dsp: warn: FW ABI is more recent than kernel
[  166.506546] sof-audio-of 556e8000.dsp: Topology: ABI 3:18:1 Kernel ABI 3:17:0
[  166.515821] sof-audio-of 556e8000.dsp: warn: topology ABI is more recent than kernel
[  166.526795] sof-audio-of 556e8000.dsp: tplg: config SAI1 fmt 0x1 mclk 12288000 width 32 slots 2 mclk id 0
[  166.537294] sof-audio-of 556e8000.dsp: ASoC: Parent card not yet available, widget card binding deferred
root@imx8qmmek:~# 
root@imx8qmmek:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: btscoaudio [bt-sco-audio], device 0: 59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0 [59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sofwm8960audio [sof-wm8960-audio], device 0: Port0 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
root@imx8qmmek:~# arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: btscoaudio [bt-sco-audio], device 0: 59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0 [59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sofwm8960audio [sof-wm8960-audio], device 0: Port0 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
Pull requests were also opened on:
- SOF repo - see [#4217](https://github.com/thesofproject/sof/pull/4217)
- xtensa_hal - see [#12](https://github.com/zephyrproject-rtos/hal_xtensa/pull/12)

cc: @dbaluta @lyakh @kv2019i @andyross @lgirdwood @MaureenHelm @dleach02

iuliana-prodan committed:
boards: xtensa: adsp: add support for imx8 board

Add support for i.MX8QM board by defining defconfig,
board configuration and device tree.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
samples/subsys/audio/sof: add configuration for imx board

Add support for i.MX8, which represents i.MX8QM board.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
dts: xtensa: add device tree for imx8

Add dtsi file for i.MX8QM and i.MX8QXP boards.
These two have the same board-level definitions,
so we call it, generically, imx8.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
cmake: add support for NXP i.MX

Xtensa GCC needs a different toolchain per SOC,
so add support for i.MX8.
If SOC_SERIES is imx8, that's for i.MX8QM and
i.MX8QXP.
If SOC_SERIES is imx8m, that's for i.MX8MP.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
soc: xtensa: adsp: add support for NXP ADSP

Add a common part for all i.MX boards.

Add support for i.MX8, which represents i.MX8QM.
This has a 1 Xtensa HiFi4 core, with 64 KB TCM,
448 KB OCRAM, 8MB SDRAM and 1 ESAI, 1 SAI as
audio interfaces.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

dbaluta commented:
Cc: @MaureenHelm

lgirdwood reviewed (approved):
Not an expert on the i.MX IP, but it looks similar to the Intel SOF Zephyr support hence LGTM.

lgirdwood commented:
ping

iuliana-prodan committed:
arch: xtensa: modify asm for interrupt sections

For IMX, for timer interrupt, the interrupt handler
was not the correct one executed and that’s because
the handlers were not at the expected address.
For IMX the size constraint of the interrupt vector
table entry is 0x1C bytes of code, less than usual.

I've added a small indirection to bypass this size
constraint and moved the default handlers to the end
of vector table, renaming them to
_Level\LVL\()VectorHelper.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

lgirdwood commented:
@MaureenHelm I think this is probably blocking on your inputs, the SOF parts and SOF PR are good. Fwiw, we aim to get this in for SOF v1.9 (next release - as this means we will be down to only 2 platforms that use xtos)

iuliana-prodan commented:
@lgirdwood @MaureenHelm, I believe we shouldn't merge this part, yet.
It's still work in progress - I'm debugging the boot failure – the memory from .noinit section (that’s for stack variables) is corrupted.
The hang-up of Zephyr looks related with multithreading mode initialization.

lgirdwood commented:
> @lgirdwood @MaureenHelm, I believe we shouldn't merge this part, yet.
> It's still work in progress - I'm debugging the boot failure – the memory from .noinit section (that’s for stack variables) is corrupted.
> The hang-up of Zephyr looks related with multithreading mode initialization.

Thanks for the update, I think @kv2019i and @andyross are also looking at linker section issues on the Intel side, these could all be related. Fwiw, the component init section is also not working, you can see it's currently hard coded in sof/zephyr/wrapper.c as a work around.

iuliana-prodan commented:
@andyross @kv2019i @lyakh @dbaluta @lgirdwood @MaureenHelm 
This is my analysis of the boot failure of SOF with Zephyr on imx.

I've debugged using `/dev/mem` and I realized that the code execution goes to` init.c, z_cstart()` function.
Next, because CONFIG_MULTITHREADING is enabled, it executes` switch_to_main_thread()` and goes to `bg_thread_main()`.
Here, it runs until` z_sys_init_run_level(_SYS_INIT_LEVEL_POST_KERNEL)`.

It should execute the POST_KERNEL functions, but fails to do so.
` enable_logger() `from log_core.c is run
 but 
 `k_sys_work_q_init()` from system_work_q.c it's not.
 
The execution flow stops between these two: enable logger and enable work_q.
The logger thread is created, but never started.

On the other hand, the interrupt handler (`_xtensa_handle_one_int3()`) is executed for a about ~100 times since the timer (`xtensa_sys_timer/sys_clock_driver_init()`) is enabled (see below my_int_mask_3_cnt variable from _xtensa_handle_one_int3()) and then it get's stuck, as well.

I see that the interrupt handler (`_xtensa_handle_one_int3()`) is executed after the logger is enabled. 
```
Code Flow                                           No. of times the intterupt handler is executed
_start() ->                                                        0
---------- z_cstart() ->                                           0
            timer enable                                           0
            multithread enable                                     0
------------------------ bg_thread_main() ->                       0
                        logger enable                              0
                        work_q enable                              0
                          logger started                           0
                          work_q started                           0
```

I've added several counters for debugging where the execution flow goes between the threads and the interrupt handlers: see my debug commit - https://github.com/zephyrproject-rtos/zephyr/pull/35467/commits/ea08a79fb2974dad505e9b7274b95bbd5ca2cd62

For debugging the bg_thread_main, I'm using 3 variables in 3 different memory sections (to check when one it gets corrupted) and I set these variables while the code is executed and I read them later, from kernel with /dev/mem.
```
__attribute__((section(".noinit"))) sof_status1[3]
__attribute__((section(".data"))) sof_status2[3]
__attribute__((section(".heap_mem"))) sof_status3[3]
```
While using these variables I've realized that after `prepare_multithreading()` is executed (so starting from `switch_to_main_thread(prepare_multithreading());` from init.c) the` .noinit `section gets corrupted - when reading the above 3 variables sof_status1 is always out of order:
```
root@imx8qmmek:~# ./readVar.sh 
==============================================================
Read sof_status
===============================
E
Reading 0x3 count starting at address 0x92429E80
0x92429E80:  00000000 9242FE60 92429FDC - sof_status1 variable
E
Reading 0x3 count starting at address 0x92430340 - sof_status2 variable
0x92430340:  33330002 33330002 33330002
E
Reading 0x3 count starting at address 0x933F4000 - sof_status3 variable
0x933F4000:  33330002 33330002 33330002
```

So, any suggestions, comments, ideas to try, are welcomed.
Here's my log - 

Thanks!

andyross commented on arch/xtensa/include/xtensa-asm2-s.h:
```diff
@@ -356,6 +356,11 @@ _restore_\@:
+#if defined(CONFIG_IMX) && (MEM_VECT_TEXT_SIZE <= 0x1C)
+.pushsection .iram.text, "ax"
+.global _Level\LVL\()VectorHelper
+_Level\LVL\()VectorHelper:
```
This seems a little suspicious.  What's the problem being solved here?  Maybe you don't have enough space in the vector table?  FWIW: there are some thought about how to better generate the entry points that might help with that.

iuliana-prodan replied:
For IMX the size constraint of the interrupt vector table entry is 0x1C bytes of code, less than usual.
I've added this indirection to bypass this size constraint and moved the default handlers to the end of vector table,  renaming them to _Level\LVL\()VectorHelper.

For the entry point I've kept the exact one we have in SOF with XTOS - _ResetVector in XTOS or _start (for Zephyr) - see https://github.com/thesofproject/sof/blob/main/src/platform/imx8/imx8.x.in#L136

andyross commented on drivers/timer/xtensa_sys_timer.c:
```diff
@@ -56,13 +56,23 @@ static void ccompare_isr(const void *arg)
 }
 
+uint32_t xtensa_sys_timer_enabled_mask_3_cnt = 0;
+extern uint32_t my_int_mask_3_cnt;
```
Debug code?

iuliana-prodan replied:
Yes, I've added a patch for debugging purposes - https://github.com/zephyrproject-rtos/zephyr/pull/35467/commits/ea08a79fb2974dad505e9b7274b95bbd5ca2cd62
I'm debugging the boot failure of SOF with Zephyr and I'm kind of stuck :(
I've explained more here: https://github.com/zephyrproject-rtos/zephyr/pull/35467#issuecomment-860954897

andyross commented on kernel/init.c:
```diff
@@ -40,6 +40,51 @@ LOG_MODULE_REGISTER(os, CONFIG_KERNEL_LOG_LEVEL);
+volatile uint32_t __aligned(64) __attribute__((section(".data"))) sof_status2[3] = {0x22221111, 0x22222222, 0x22223333};
+volatile uint32_t __aligned(64) __attribute__((section(".heap_mem"))) sof_status3[3] = {0x33331111, 0x33332222, 0x33333333};
+
+#ifdef CONFIG_IMX
```
Obvoiusly none of this code belongs here.  If you want SOC-specific initialization hooks, they should be called out of the arch/soc layer before entry to z_cstart(), in arch_kernel_init() (called just after entry to z_cstart()), or via a standard device struct initialized via SYS_INIT

iuliana-prodan replied:
This code is just for debugging - see commit https://github.com/zephyrproject-rtos/zephyr/pull/35467/commits/ea08a79fb2974dad505e9b7274b95bbd5ca2cd62

I've added several counters and status variables and I read them, from Linux, with /dev/mem.

andyross commented on kernel/init.c:
```diff
@@ -190,39 +235,49 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
+	 *
+	 * Verify how many times the interrupt handler was executed so far
+	 */
+	init_bg_thread_main_mask_3_cnt = my_int_mask_3_cnt;
```
More debug code?

iuliana-prodan replied:
Yes, more debug code, part of [do not merge] commit https://github.com/zephyrproject-rtos/zephyr/pull/35467/commits/ea08a79fb2974dad505e9b7274b95bbd5ca2cd62

andyross commented on kernel/work.c:
```diff
@@ -18,6 +18,33 @@
+
+void my_cache_flush(void *addr, size_t bytes);
+
+void set_work_q_status(uint32_t status)
```
What's this for?

iuliana-prodan replied:
Also debug code.

andyross commented on soc/xtensa/nxp_adsp/common/interrupt.c:
```diff
@@ -0,0 +1,490 @@
+/*
+ * The IRQ_STEER module takes 512 shared interrupts and delivers them
+ * via 8 interrupt lines to any other component. It supports 5 channels,
+ * one of them being for the DSP itself (channel 0).
```
You might want to consider moving this code into drivers/interrupt_controller.  That's the way other SOCs do it, though I don't have strong feelings (an interrupt controller is never going to be a "normal" device, it's very low level soc/arch code by definition).

iuliana-prodan replied:
I've removed it.
We'll use the irq_steer from SOF, for now.

andyross commented on soc/xtensa/nxp_adsp/imx8/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,282 @@
+
+	my_cache_flush(sof_status1, sizeof(sof_status1));
+	my_cache_flush(sof_status2, sizeof(sof_status2));
+	my_cache_flush(sof_status3, sizeof(sof_status3));
```
I'm guessing again that this is debug code?  Obviously that warning about about automatically generated code is being violated otherwise (though we never integrated that python script into the proper build and IIRC it's been accidentally removed from the tree a few times... really Xtensa interrupt entry needs a rewrite)

iuliana-prodan replied:
Yes, also debug code.

iuliana-prodan replied:
I've removed all the debug code.

andyross commented on soc/xtensa/nxp_adsp/imx8/include/soc/memory.h:
```diff
@@ -0,0 +1,251 @@
+
+#define SRAM_RESERVE_HEADER_SPACE	0x00010000
+
+#define SRAM_BASE	0x596E8400
```
SRAM_BASE is only 1k-aligned?  Is that correct?  Or is this a way of getting code out of the way of something special at the bottom of memory?  If so, calling it "base" is maybe confusing.

iuliana-prodan replied:
This is an unfortunate copy-paste.
The address is the iMX IRAM address, but is never used.

iuliana-prodan replied:
I've removed it.

andyross commented on soc/xtensa/nxp_adsp/imx8/include/soc/memory.h:
```diff
@@ -0,0 +1,251 @@
+#define SRAM_RESERVE_HEADER_SPACE	0x00010000
+
+#define SRAM_BASE	0x596E8400
+#define SRAM_SIZE	0x880000
```
Some of this looks like it might be copied over from the intel_adsp code?  Note that that code comes directly from SOF.  For new platforms, the Zephyr model is to put most intelligence like this about memory layout and register addresses into devicetree files.

iuliana-prodan replied:
These two, SRAM_BASE and SRAM_SIZE, are never used for iMX.
The address and the size are correct, but the naming is incorrect.

But, since they are not used, I'll remove them.

iuliana-prodan replied:
I've removed the unused variables from imx8, memory.h file.

andyross commented on soc/xtensa/nxp_adsp/imx8/include/soc/memory.h:
```diff
@@ -0,0 +1,251 @@
+
+#define SRAM_INBOX_BASE		(SRAM_OUTBOX_BASE + SRAM_OUTBOX_SIZE)
+#define SRAM_INBOX_SIZE		0x1000
+#define SRAM_INBOX_OFFSET	SRAM_OUTBOX_SIZE
```
More stuff copied over from SOF/intel_adsp?  The inbox/outbox stuff are specific to the communication with the host x86 CPUs.  Surely they aren't relevant to you?

iuliana-prodan replied:
We also have this on IMX for SOF with XTOS -  see https://github.com/thesofproject/sof/blob/main/src/platform/imx8/include/platform/lib/memory.h#L83

andyross reviewed (commented):
Some notes.  Nothing looks like a terrible mistake that I can see.  I do see it's managed to suck in a bunch of stale code from SOF that is likely to be removed as the Zephyr port evolves.

iuliana-prodan commented:
> Some notes. Nothing looks like a terrible mistake that I can see. I do see it's managed to suck in a bunch of stale code from SOF that is likely to be removed as the Zephyr port evolves.

@andyross thanks for reviewing!
Do you have any suggestions regarding https://github.com/zephyrproject-rtos/zephyr/pull/35467#issuecomment-860954897?

The problem I have is that the execution flow gets stuck-  I don't see any threads running. 
And the interrupt handler is executed only ~100 times.

iuliana-prodan commented:
> > Some notes. Nothing looks like a terrible mistake that I can see. I do see it's managed to suck in a bunch of stale code from SOF that is likely to be removed as the Zephyr port evolves.
> 
> @andyross thanks for reviewing!
> Do you have any suggestions regarding [#35467 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/35467#issuecomment-860954897)?
> 
> The problem I have is that the execution flow gets stuck- I don't see any threads running.
> And the interrupt handler is executed only ~100 times.

@andyross on a normal Zephyr execution flow, should we get to `xtensa_excint1_c()` from [xtensa-asm2.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/xtensa/core/xtensa-asm2.c#L174)?

MaureenHelm reviewed (dismissed):
Thanks @andyross for your review. I don't have much to add, but I'm not very familiar with Xtensa yet.

iuliana-prodan committed:
west: sign: add support for NXP i.MX

Add support for signing i.MX SOF with Zephyr images with rimage.

Note that, for i.MX, we don't need a bootloader nor the .elf.mod files
since there is no need to change the VMA/LMA of each uncached section
to the equivalent address in the cached area of memory.
For the above reasons, I've updated the arguments given to "west sign"
command.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
CODEOWNERS: add codeowner for SOF with Zephyr on NXP i.MX8

Add myself and Daniel Baluta as codeowners for i.MX8 related files for
SOF with Zephyr OS.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented on dts/xtensa/nxp/nxp_imx8.dtsi:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2021 NXP Semiconductors
```
Done

iuliana-prodan commented:
> > @lgirdwood @MaureenHelm, I believe we shouldn't merge this part, yet.
> > It's still work in progress - I'm debugging the boot failure – the memory from .noinit section (that’s for stack variables) is corrupted.
> > The hang-up of Zephyr looks related with multithreading mode initialization.
> 
> Thanks for the update, I think @kv2019i and @andyross are also looking at linker section issues on the Intel side, these could all be related. Fwiw, the component init section is also not working, you can see it's currently hard coded in sof/zephyr/wrapper.c as a work around.

@lgirdwood @MaureenHelm @andyross @kv2019i @dbaluta this is now ready for review and it can be merged, if it's ok with you.
As mentioned in the pull request I obtain now the zephyr.ri image which boots the DSP and loads the topology.

@andyross I've also removed the debug code :)

dbaluta commented:
@iuliana-prodan  this looks good to me AFAICT. @MaureenHelm can you help review this?

MaureenHelm commented:
> @iuliana-prodan this looks good to me AFAICT. @MaureenHelm can you help review this?

@iuliana-prodan @dbaluta Can you have a look at the buildkite failure?
```
ERROR   - Cmake build failure: /workdir/zephyr/samples/subsys/audio/sof for nxp_adsp_imx8
INFO    -  11/123 nxp_adsp_imx8             samples/subsys/audio/sof/sample.audio.sof          FAILED Cmake build failure (build)
INFO    - ---/workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/build.log----
ERROR   - Including boilerplate (Zephyr base): /workdir/zephyr/cmake/app/boilerplate.cmake
-- Application: /workdir/zephyr/samples/subsys/audio/sof
-- Zephyr version: 2.6.99 (/workdir/zephyr), build: zephyr-v2.6.0-744-g14e65a9e1e68
-- Found Python3: /usr/bin/python3.8 (found suitable exact version "3.8.5") found components: Interpreter
-- Found west (found suitable version "0.11.0", minimum required is "0.7.1")
-- Board: nxp_adsp_imx8
-- Cache files will be written to: /root/.cache/zephyr
-- Using toolchain: zephyr 0.12.4 (/opt/toolchains/zephyr-sdk-0.12.4)
-- Found dtc: /opt/toolchains/zephyr-sdk-0.12.4/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /workdir/zephyr/boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8.dts
-- Generated zephyr.dts: /workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/zephyr/include/generated/device_extern.h
 
warning: SMP_BOOT_DELAY (defined at kernel/Kconfig:793) was assigned the value 'y' but got the value
'n'. Check these unsatisfied dependencies: SMP (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_SMP_BOOT_DELAY.html and/or look up
SMP_BOOT_DELAY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.
 
Parsing /workdir/zephyr/Kconfig
Loaded configuration '/workdir/zephyr/boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8_defconfig'
Merged configuration '/workdir/zephyr/samples/subsys/audio/sof/prj.conf'
Merged configuration '/workdir/zephyr/samples/subsys/audio/sof/boards/nxp_adsp_imx8.conf'
Configuration saved to '/workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/zephyr/.config'
Kconfig header saved to '/workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/toolchains/zephyr-sdk-0.12.4/xtensa/nxp_imx_adsp/xtensa-zephyr-elf/bin/xtensa-zephyr-elf-gcc
CMake Error at /var/lib/buildkite-agent/zephyr-module-cache/modules/audio/sof/zephyr/CMakeLists.txt:332 (zephyr_library_sources):
  zephyr_library_sources Function invoked with incorrect arguments for
  function named: zephyr_library_sources
 
-- Found Python3: /usr/bin/python3.8 (found version "3.8.5") found components: Interpreter
-- version.cmake starting SOF build at 2021-07-06T09:46:09Z UTC
-- Building git commit with parent(s):
14e65a9e1e 4265955ba1 (HEAD) CODEOWNERS: add codeowner for SOF with Zephyr on NXP i.MX8
-- GIT_TAG / GIT_LOG_HASH : v1.5-rc1-1471-g779f28ed465c / 779f28ed4
CMake Warning at /var/lib/buildkite-agent/zephyr-module-cache/modules/audio/sof/scripts/cmake/version.cmake:116 (message):
  Source content hash not computed without git, using GIT_LOG_HASH instead
Call Stack (most recent call first):
  /var/lib/buildkite-agent/zephyr-module-cache/modules/audio/sof/zephyr/CMakeLists.txt:543 (include)
 
-- Generating /var/lib/buildkite-agent/zephyr-module-cache/modules/audio/sof/zephyr/include/version.h
-- Configuring incomplete, errors occurred!
See also "/workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/CMakeFiles/CMakeOutput.log".
See also "/workdir/zephyr/twister-out/nxp_adsp_imx8/samples/subsys/audio/sof/sample.audio.sof/CMakeFiles/CMakeError.log".
```

iuliana-prodan commented:
> > @iuliana-prodan this looks good to me AFAICT. @MaureenHelm can you help review this?
> 
> @iuliana-prodan @dbaluta Can you have a look at the buildkite failure?
> 

I don't know why is this failing.
Can you please help me understand how buildkite is running?

I'm building and signing the image using the following commands:
```
west build -d imx -b nxp_adsp_imx8 zephyr/samples/subsys/audio/sof/
west sign -d imx -p rimage/build/rimage -t rimage -D rimage/config -- -k  no
```
I've started from [here](https://docs.zephyrproject.org/latest/guides/beyond-GSG.html#build-and-run-an-application) and [here](https://thesofproject.github.io/latest/getting_started/build-guide/build-with-zephyr.html#).

Here's my build log:
```
nxa06898@lsv15040:/opt/samba/nxa06898/zephyrproject$ west build -d imx -b nxp_adsp_imx8 zephyr/samples/subsys/audio/sof/
-- west build: generating a build system
Including boilerplate (Zephyr base): /opt/samba/nxa06898/zephyrproject/zephyr/cmake/app/boilerplate.cmake
-- Application: /opt/samba/nxa06898/zephyrproject/zephyr/samples/subsys/audio/sof
-- Zephyr version: 2.6.99 (/opt/samba/nxa06898/zephyrproject/zephyr), build: zephyr-v2.6.0-744-g14e65a9e1e68
-- Found Python3: /usr/bin/python3.6 (found suitable exact version "3.6.9") found components: Interpreter 
-- Found west (found suitable version "0.11.0", minimum required is "0.7.1")
-- Board: nxp_adsp_imx8
-- Cache files will be written to: /home/nxa06898/.cache/zephyr
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Using toolchain: zephyr 0.12.2 (/opt/zephyr-sdk-0.12.2)
-- Found dtc: /opt/zephyr-sdk-0.12.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /opt/samba/nxa06898/zephyrproject/zephyr/boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8.dts
-- Generated zephyr.dts: /opt/samba/nxa06898/zephyrproject/imx/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /opt/samba/nxa06898/zephyrproject/imx/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /opt/samba/nxa06898/zephyrproject/imx/zephyr/include/generated/device_extern.h
Parsing /opt/samba/nxa06898/zephyrproject/zephyr/Kconfig
Loaded configuration '/opt/samba/nxa06898/zephyrproject/zephyr/boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8_defconfig'
Merged configuration '/opt/samba/nxa06898/zephyrproject/zephyr/samples/subsys/audio/sof/prj.conf'
Merged configuration '/opt/samba/nxa06898/zephyrproject/zephyr/samples/subsys/audio/sof/boards/nxp_adsp_imx8.conf'
Configuration saved to '/opt/samba/nxa06898/zephyrproject/imx/zephyr/.config'
Kconfig header saved to '/opt/samba/nxa06898/zephyrproject/imx/zephyr/include/generated/autoconf.h'

warning: SMP_BOOT_DELAY (defined at kernel/Kconfig:793) was assigned the value 'y' but got the value
'n'. Check these unsatisfied dependencies: SMP (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_SMP_BOOT_DELAY.html and/or look up
SMP_BOOT_DELAY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/zephyr-sdk-0.12.2/xtensa/nxp_imx_adsp/xtensa-zephyr-elf/bin/xtensa-zephyr-elf-gcc
-- Found Python3: /usr/bin/python3.6 (found version "3.6.9") found components: Interpreter 
-- version.cmake starting SOF build at 2021-07-07T09:20:08Z UTC
-- Building git commit with parent(s):
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
-- GIT_TAG / GIT_LOG_HASH : v1.8-rc1-302-gcf91943acfb3 / cf91943ac
CMake Warning at /opt/samba/nxa06898/sof/sof/scripts/cmake/version.cmake:116 (message):
  Source content hash not computed without git, using GIT_LOG_HASH instead
Call Stack (most recent call first):
  /opt/samba/nxa06898/sof/sof/zephyr/CMakeLists.txt:576 (include)


-- Up-to-date /opt/samba/nxa06898/sof/sof/zephyr/include/version.h
-- Configuring done
-- Generating done
-- Build files have been written to: /opt/samba/nxa06898/zephyrproject/imx
-- west build: building application
[1/224] Preparing syscall dependency handling

[217/224] Linking C executable zephyr/zephyr_prebuilt.elf

[224/224] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
vector_reset_text:         289 B        736 B     39.27%
vector_reset_lit:          0 GB        288 B      0.00%
vector_base_text:         362 B        376 B     96.28%
 vector_int2_lit:          0 GB          4 B      0.00%
vector_int2_text:           3 B         28 B     10.71%
 vector_int3_lit:          0 GB          4 B      0.00%
vector_int3_text:           3 B         28 B     10.71%
 vector_int4_lit:          0 GB          4 B      0.00%
vector_int4_text:           3 B         28 B     10.71%
 vector_int5_lit:          0 GB          4 B      0.00%
vector_int5_text:           3 B         28 B     10.71%
vector_kernel_lit:          0 GB          4 B      0.00%
vector_kernel_text:           3 B         28 B     10.71%
 vector_user_lit:          0 GB          4 B      0.00%
vector_user_text:          18 B         28 B     64.29%
vector_double_lit:          0 GB          4 B      0.00%
vector_double_text:           6 B         28 B     21.43%
 iram_text_start:         192 B        424 B     45.28%
          sdram0:      202016 B         8 MB      2.41%
          sdram1:       8144 KB      8172 KB     99.66%
        IDT_LIST:          0 GB         8 KB      0.00%
static_uuid_entries_seg:        1104 B        24 KB      4.49%
static_log_entries_seg:          0 GB        32 MB      0.00%
 fw_metadata_seg:         720 B        32 MB      0.00%
```
Here's the complete log, for build and sign: 
[west_build_and_sign_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/6776028/west_build_and_sign_log.txt)

The failure might be due to the SOF module brought, by default, to Zephyr.

In my environment, I'm using a symbolic link to the SOF module, where I've added the support for i.MX with Zephyr (see pull request [here](https://github.com/thesofproject/sof/pull/4217) - this was merged on SOF master branch).
@nashif should I cherry-pick my SOF pull request ([#4217](https://github.com/thesofproject/sof/pull/4217)) on zephyr branch from[ zephyrproject-rtos/sof](https://github.com/zephyrproject-rtos/sofl) repo?

cc: @MaureenHelm @nashif @andyross @dbaluta @lgirdwood

lyakh commented on cmake/toolchain/zephyr/0.11/target.cmake:
```diff
@@ -27,6 +27,10 @@ if("${ARCH}" STREQUAL "xtensa")
   elseif("${SOC_SERIES}" STREQUAL "broadwell_adsp")
     set(SR_XT_TC_SOC intel_bdw_adsp)
+  elseif("${SOC_SERIES}" STREQUAL "imx8")
+    set(SR_XT_TC_SOC nxp_imx_adsp)
```
why isn't it `nxp_imx8_adsp`?

iuliana-prodan replied:
The `zephyr-sdk-0.12.2/xtensa/` contains 2 toolchains: `nxp_imx_adsp` for imx8QM and imx8QXP and `nxp_imx8m_adsp` for imx8MP. 
So, the corresponding toolchain for `imx8` (which represents imx8QM and imx8QXP) is `nxp_imx_adsp`.

dbaluta replied:
That is my fault. In the past when I added the toolchain for i.MX8QM i named it imx instead of imx8. To make things 'better' the toolchain for imx8M Plus is indeed named imx8m.

MaureenHelm commented:
> The failure might be due to the SOF module brought, by default, to Zephyr.
> 
> In my environment, I'm using a symbolic link to the SOF module, where I've added the support for i.MX with Zephyr (see pull request [here](https://github.com/thesofproject/sof/pull/4217) - this was merged on SOF `main `branch).
> @nashif should I cherry-pick my SOF pull request ([#4217](https://github.com/thesofproject/sof/pull/4217)) on `zephyr `branch from[ zephyrproject-rtos/sof](https://github.com/zephyrproject-rtos/sofl) repo?

Yes, the SOF module needs to be updated. See [Submitting changes to modules](https://docs.zephyrproject.org/latest/guides/modules.html#submitting-changes-to-modules) for details about how to do this.

iuliana-prodan commented:
@nashif @lgirdwood can you please help me update the SOF module from Zephyr with the i.MX support?

I see that the SOF module brought to zephyrproject-rtos is a [fork ](https://github.com/zephyrproject-rtos/sof)from thesofproject/sof so how can I update the fork:
1. cherry pick only the i.MX support from thesofproject/sof, main branch;
2. update to latest main commit;
Please let me know which of the above solution should I use, or maybe you think of something else.

Next, I will update the commit id from[ west.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml#L197) file for sof module.

Thank you!

cc: @lyakh @kv2019i @marc-hb @dbaluta @MaureenHelm @andyross

lgirdwood commented:
> @nashif @lgirdwood can you please help me update the SOF module from Zephyr with the i.MX support?
> 
> I see that the SOF module brought to zephyrproject-rtos is a [fork ](https://github.com/zephyrproject-rtos/sof)from thesofproject/sof so how can I update the fork:

It's not a fork - just a clone.

> 
> 1. cherry pick only the i.MX support from thesofproject/sof, main branch;
> 2. update to latest main commit;

Yes option 2,  I have just pushed ```main``` HEAD from now. It turns out the The Zephyr clone repo was using the old ```master``` naming so please also update the manifest to use the main branch instead of master.

@nashif can we add @dbaluta and @iuliana-prodan to have push access to the Zephyr SOF module. This will allow us to split the work and can both update as Intel/NXP release schedules dictate.

iuliana-prodan commented:
I've created pull requests to update `xtensa_hal` and `SOF module` with the imx8 audio DSP support:
1. https://github.com/zephyrproject-rtos/zephyr/pull/37209
2. https://github.com/zephyrproject-rtos/zephyr/pull/37250

dbaluta commented:
@iuliana-prodan can you rebase this and resubmit in order to trigger the CI again. The build problem should be fixed now, right?

iuliana-prodan committed:
manifest: update sof module

Update SOF module to latest main to fix
SOF with Zephyr on i.MX.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
Changes from pervious version:
- rebase to latest main;
- update `west.yaml`, for SOF module with the sha-1 of latest commit, instead of pull request number.

dbaluta commented:
Thanks @iuliana-prodan. Awesome work!
